default tar filename has extra dash in custom metric mode

karishmalalani opened the pull request:
Issue fixed extra dash has been removed

karishmalalani committed:
default tar filename has extra dash in custom metric mode

Issue fixed extra dash has been removed

karishmalalani committed:
Merge branch 'master' into file-archive

# Conflicts:
#	promdump/promdump.go

ionthegeek commented on promdump/promdump.go:
```diff
@@ -32,9 +31,11 @@ import (
 	"time"
 
+	ywclient "github.com/yugabyte/platform-go-client"
+
 	"github.com/dsnet/compress/bzip2"
 	"github.com/prometheus/client_golang/api"
-	"github.com/prometheus/client_golang/api/prometheus/v1"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
```
Avoid making spurious changes like this.

karishmalalani replied:
This time I commit with v1

ionthegeek replied:
Please revert this change and also the change above that moves the `ywclient` line.

ionthegeek commented on promdump/promdump.go:
```diff
@@ -641,15 +642,39 @@ func addToArchive(tw *tar.Writer, filename string) error {
+	// If node_prefix is not found in metric, use nodePrefix flag
+	if prefix == "" && *nodePrefix != "" {
+		prefix = *nodePrefix
+	}
```
We don't need to be this clever. Just test `if *nodePrefix == ""` to figure out whether a node prefix has been provided or not. We don't need to handle the case where somebody has put `node_prefix` into the custom metric.

karishmalalani replied:
As per your suggestion, the code now only looks for the flag "nodePrefix" and does not find "nodePrefix" for other cases.

ionthegeek commented on promdump/promdump.go:
```diff
@@ -641,15 +642,39 @@ func addToArchive(tw *tar.Writer, filename string) error {
+	// If neither is available, use a default prefix
+	if prefix == "" {
+		return fmt.Sprintf("promdump-%s.tar", time.Now().Format("20060102-150405"))
+	}
```
This doesn't handle the case where a custom metric is specified and compression is enabled.

You'll probably want to switch to a string builder kind of approach where you put the filename together in pieces by appending, then return the assembled string at the end of the func. Concatenate `promdump-` with the node prefix (if provided), the timestamp string, and the appropriate extension(s).

karishmalalani replied:
I did not use string builder but change logic which do similar like string builder

ionthegeek reviewed (changes requested):
This seems unnecessarily complex. There is a bug in handling the case where compression is enabled for a custom metric promdump.

Feedback inline.

karishmalalani committed:
Merge branch 'master' into file-archive

karishmalalani committed:
default tar filename has extra dash in custom metric mode

Make changes as per feedback

now code give default name for custom metric without extra '-'

karishmalalani committed:
Revert "default tar filename has extra dash in custom metric mode"

This reverts commit 5be619b8254ad748cdf8ff239e82d8537858d4bf.

karishmalalani committed:
Filename has extra dash in custom metric mode

Changes were made as per feedback

ionthegeek commented on promdump/promdump.go:
```diff
@@ -645,15 +646,30 @@ func addToArchive(tw *tar.Writer, filename string) error {
 
-	return fmt.Sprintf("promdump-%s-%s.tar", *nodePrefix, now().Format("20060102-150405"))
+	// Default to ".tar" if no compression type is specified
+	return filename + ".tar"
```
This approach results in a bunch of memory copying, which is inefficient. Because strings are immutable, we have to allocate a new string object every time we append or concatenate strings in this way.

Use strings.Builder instead:
https://pkg.go.dev/strings#example-Builder

```
var filename strings.Builder

filename.WriteString("promdump")

// Check if nodePrefix is provided and append it to the filename if not empty
if *nodePrefix != "" {
  fmt.Fprintf(&filename, "-%s", *nodePrefix)
}

// Append the timestamp
fmt.Fprintf(&filename, "-%s", time.Now().Format("20060102-150405"))

...

return filename.String()
```

ionthegeek commented on promdump/promdump.go:
```diff
@@ -645,15 +646,30 @@ func addToArchive(tw *tar.Writer, filename string) error {
+	filename = "promdump"
+
+	// Check if nodePrefix is provided and append it to the filename if not empty
+	if nodePrefix != nil && *nodePrefix != "" {
```
The `nodePrefix != nil` portion of this conditional is spurious. This condition will always be true since it's a pointer to a (possibly nil) string. Simplify this to `if *nodePrefix != "" {`.

ionthegeek commented on promdump/promdump.go:
```diff
@@ -1402,6 +1418,14 @@ func main() {
 					}
 				}
 			}
+			if *out != "" {
+
+				_, err := cleanFiles(*out, customMetricCount, false)
+				if err != nil {
+					log.Printf("Error cleaning files : %v", err)
+				}
+
+			}
```
Please keep changes to one issue per PR.

ionthegeek reviewed (changes requested):
Needs some changes. Mostly janitorial, but one major one.

karishmalalani committed:
Use strings.Builder

As per feedback generateDefaultTarFilename() use strings.builder

ionthegeek commented on promdump/promdump.go:
```diff
@@ -645,15 +646,33 @@ func addToArchive(tw *tar.Writer, filename string) error {
+	// Append the timestamp
+	fmt.Fprintf(&filename, "-%s", time.Now().Format("20060102-150405"))
+
+	// file extension based on compression type
+	if tarCompression != nil {
+		switch *tarCompression {
+		case "gzip":
+			return filename.String() + ".tar.gz"
+		case "bzip2":
+			return filename.String() + ".tar.bz2"
+		}
+	}
+
+	// Default to ".tar" if no compression type is specified
+	return filename.String() + ".tar"
+
```
This is section is still doing unnecessary memory copying. If you're returning `StringBuilder + something`, you're probably doing an unnecessary memory copy.

The program flow could be made clearer by appending the extensions like this:

```
filename.WriteString(".tar")

if tarCompression != nil {
  switch *tarCompression {
  case "gzip":
    filename.WriteString(".gz")
  case "bzip2":
    filename.WriteString(".bz2")
  }
}

return filename.String()
```

ionthegeek commented on promdump/promdump.go:
```diff
@@ -15,7 +15,6 @@ import (
 	"errors"
 	"flag"
 	"fmt"
-	ywclient "github.com/yugabyte/platform-go-client"
```
Revert this removal

ionthegeek commented on promdump/promdump.go:
```diff
@@ -32,6 +31,8 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/yugabyte/platform-go-client"
+
```
Revert this addition

ionthegeek reviewed (changes requested):
Avoid making spurious changes like moving libraries around in the `import` section. Use `git diff` to review what's changed and if you see something that your IDE or code formatter has changed for no reason, revert that change before submitting.

The handling of file extensions could still be improved. I've provided a specific change in the comments.

There is a section added for cleaning custom metric files. Split this out into a separate issue and PR.
